### PR TITLE
MC-30626: Custom Option with Percent price is converted to a currency twice

### LINKED ISSUE #26432: Custom Option with Percent price is converted to a currency twice
### Preconditions (*)

1. Magento 2.4-develop
2. PHP 7.2

### Steps to reproduce (*)

1. Add custom option of type Field, Radio, Checkbox or Date to a product
2. Set price to be of type percent
3. Add another currency to a shop
4. On a product page change a currency to different from shops base currency so that magento converts the prices.

### Expected result (*)

1. Price of custom option should be a percent of products price but in current currency.

### Actual result (*)

1. Percent price is converted twice.

Magento\Catalog\Pricing\Price\CustomOptionPriceCalculator::getOptionPriceByPriceCode():
in case of a percent, is getting price from product's PriceInfo which is converted to a selected currency.
However then on a block levels there different methods for different types of custom options, that end up in CustomOptionPrice::getCustomAmount() that does convertAndRound on already converted amount in case of percent custom option price.

 ### Additional information
-detailed steps with screens https://github.com/magento/magento2/issues/26432#issuecomment-649813348

### PR DESCRIPTION
**Pull request by engcom-Echo**
### Fixed Issues (if relevant)

1. Fixes magento/magento2#26432

### Manual testing scenarios (*)

See: magento/magento2#26432

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

File: app/code/Magento/Catalog/Pricing/Price/CustomOptionPrice.php [hunk 1]
```diff
@@ -161,6 +163,19 @@ public function getCustomOptionRange($getMin, $priceCode = \Magento\Catalog\Pric
...
     }
 
+    /**
+     * Get custom amount without currency convert
+     *
+     * @param float|string $amount
+     * @param null|array $context
+     * @return AmountInterface
+     */
+    public function getCustomRoundAmount($amount, $context = [])
```
Comment (engcom-Kilo) [inline app/code/Magento/Catalog/Pricing/Price/CustomOptionPrice.php]: There is no getCustomRoundAmount in CustomOptionPriceInterface. Please consider to find other solution.

File: app/code/Magento/Catalog/Test/Mftf/Test/StorefrontCheckCustomOptionPriceDifferentCurrencyTest.xml [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        </before>
+        <after>
+            <magentoCLI command="config:set currency/options/allow USD" stepKey="setCurrencyAllow"/>
+            <deleteData createDataKey="createProduct" stepKey="deleteProduct"/>
+            <deleteData createDataKey="createCategory" stepKey="deleteCategory"/>
+        </after>
+        <actionGroup ref="StorefrontOpenProductEntityPageActionGroup" stepKey="openProductPageOnStorefront">
+            <argument name="product" value="$createProduct$"/>
+        </actionGroup>
+        <seeElement selector="{{StorefrontProductInfoMainSection.productAttributeOptionsCheckbox(ProductOptionCheckbox.title, '12.3')}}" stepKey="checkPriceProductOptionUSD"/>
```
Comment (engcom-Kilo) [inline app/code/Magento/Catalog/Test/Mftf/Test/StorefrontCheckCustomOptionPriceDifferentCurrencyTest.xml]: Please create ActionGroup for this action.

**Review by engcom-Kilo — CHANGES_REQUESTED** [review]

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-8475](https://jira.corp.magento.com/browse/ENGCOM-8475) has been created to process this Pull Request
:eight_spoked_asterisk: @ihor-sviziev, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Review by gabrieldagama — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @gabrieldagama, thank you for the review.
[ENGCOM-8475](https://jira.corp.magento.com/browse/ENGCOM-8475) has been created to process this Pull Request

File: app/code/Magento/Catalog/Test/Mftf/Test/StorefrontCheckCustomOptionPriceDifferentCurrencyTest.xml [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+            <stories value="Custom options"/>
+            <title value="Check custom option price with different currency"/>
+            <description value="Check custom option price with different currency on the product page"/>
+            <severity value="CRITICAL"/>
+            <testCaseId value="MC-38926"/>
+            <useCaseId value="MC-30626"/>
+            <group value="catalog"/>
+        </annotations>
+        <before>
+            <magentoCLI command="config:set currency/options/allow EUR,USD" stepKey="setCurrencyAllow"/>
```
Comment (viktorpetryk) [inline app/code/Magento/Catalog/Test/Mftf/Test/StorefrontCheckCustomOptionPriceDifferentCurrencyTest.xml]: Could you use the existing entity for this?
```
<magentoCLI command="config:set {{SetAllowedCurrenciesConfigForUSD.path}} {{SetAllowedCurrenciesConfigForUSD.value}},{{SetAllowedCurrenciesConfigForEUR.value}}" stepKey="setAllowedCurency"/>

File: app/code/Magento/Catalog/Test/Mftf/Test/StorefrontCheckCustomOptionPriceDifferentCurrencyTest.xml [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+                <requiredEntity createDataKey="createCategory"/>
+                <field key="price">10</field>
+            </createData>
+            <updateData createDataKey="createProduct" entity="productWithCheckbox" stepKey="updateProductWithOptions"/>
+            <actionGroup ref="CliIndexerReindexActionGroup" stepKey="reindex">
+                <argument name="indices" value=""/>
+            </actionGroup>
+        </before>
+        <after>
+            <magentoCLI command="config:set currency/options/allow USD" stepKey="setCurrencyAllow"/>
```
Comment (viktorpetryk) [inline app/code/Magento/Catalog/Test/Mftf/Test/StorefrontCheckCustomOptionPriceDifferentCurrencyTest.xml]: Same here:
```
<magentoCLI command="config:set {{SetAllowedCurrenciesConfigForUSD.path}} {{SetAllowedCurrenciesConfigForUSD.value}}" stepKey="setAllowedCurency"/>

File: app/code/Magento/Catalog/Test/Mftf/ActionGroup/StorefrontAssertCustomOptionCheckboxByPriceActionGroup.xml [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+<!--
+ /**
+  * Copyright © Magento, Inc. All rights reserved.
+  * See COPYING.txt for license details.
+  */
+-->
+
+<actionGroups xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+              xsi:noNamespaceSchemaLocation="urn:magento:mftf:Test/etc/actionGroupSchema.xsd">
+    <actionGroup name="StorefrontAssertCustomOptionCheckboxByPriceActionGroup">
```
Comment (viktorpetryk) [inline app/code/Magento/Catalog/Test/Mftf/ActionGroup/StorefrontAssertCustomOptionCheckboxByPriceActionGroup.xml]: If the action group is making an assertion, then use the following format: Assert{Admin or Storefront}{Functionality}ActionGroup.xml.

File: app/code/Magento/Catalog/Test/Mftf/ActionGroup/StorefrontAssertCustomOptionCheckboxByPriceActionGroup.xml [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+        <annotations>
+            <description>Validates that the provided price for Custom Option Checkbox is present on the Storefront Product page.</description>
+        </annotations>
+        <arguments>
+            <argument name="optionTitle" type="string" defaultValue="{{ProductOptionCheckbox.title}}"/>
+            <argument name="price" type="string"/>
```
Comment (viktorpetryk) [inline app/code/Magento/Catalog/Test/Mftf/ActionGroup/StorefrontAssertCustomOptionCheckboxByPriceActionGroup.xml]: Action group arguments SHOULD specify default values.

File: app/code/Magento/Catalog/Block/Product/View/Options/AbstractOptions.php [hunk 7]
```diff
@@ -187,8 +206,14 @@ protected function _formatPrice($value, $flag = true)
...
-        $context = [CustomOptionPriceInterface::CONFIGURATION_OPTION_FLAG => true];
-        $optionAmount = $customOptionPrice->getCustomAmount($value['pricing_value'], null, $context);
+        $context[CustomOptionPriceInterface::CONFIGURATION_OPTION_FLAG] = true;
+        $optionAmount = $isPercent
+            ? $this->calculator->getAmount(
+                $this->priceCurrency->roundPrice($value['pricing_value']),
+                $this->getProduct(),
+                null,
+                $context
+            ): $customOptionPrice->getCustomAmount($value['pricing_value'], null, $context);
```
Comment (viktorpetryk) [inline app/code/Magento/Catalog/Block/Product/View/Options/AbstractOptions.php]: Please add space between ):

**Review by viktorpetryk — CHANGES_REQUESTED** [review]

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Review by viktorpetryk — APPROVED** [review]

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-8475](https://jira.corp.magento.com/browse/ENGCOM-8475) has been created to process this Pull Request

**Comment by engcom-Bravo** [comment]
:heavy_check_mark: QA Passed

### Manual testing scenario
1. From Admin, create a _Simple Product_ with a base price of _100_, and a  **Custom option** which adds _50%_ to the base price 
cust_option
2. Go to _Stores - Configuration - General - Currency Setup_ and assert that _Euro_ is checked in **Allowed Currencies**
euro_allowed
3. Go to _Stores - Currency Rates_ and set _Rate_, for example, _2 EUR_ to _1 USD_
rates
4. Go to Storefront, and open the previously created Simple Products page 
5. Change the Stores currency to _EUR - Euro_
web_curr
6. Check the _Custom option_ calculation

**BEFORE** applying changes provided in this PR, the Custom options was calculated wrong: Base price + Custom option price + Custom option price one more time
before_currency

**AFTER** switching to the PR, the Custom option price is calculated correctly, only once
after_currency

The order can be placed successfully and all the prices are displayed correctly in order Details
order_total

**Comment by ihor-sviziev** [comment]
@engcom-Bravo seems like you missed to move this PR forward?

**Comment by engcom-Echo** [comment]
@magento run Integration Tests

**Comment by zakdma** [comment]
@magento import code to https://github.com/magento-tsg/magento2ce

**Comment by magento-engcom-team** [comment]
@zakdma the branch with code successfully imported into`magento-tsg/magento2ce` repository. Branch name: `imported-magento-magento2-30807`.

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Pull request merged**